src/Microsoft.OData.Core/ODataBatchReader.cs

```diff
@@ -204,440 +233,82 @@ void IODataBatchOperationListener.BatchOperationContentStreamDisposed()
         {
-            switch (this.batchReaderState)
+            this.currentBatchSize++;
+
+            if (this.currentBatchSize > this.inputContext.MessageReaderSettings.MessageQuotas.MaxPartsPerBatch)
             {
-                case ODataBatchReaderState.Initial:
-                    // If we find an end boundary when in state 'Initial' it means that we
-                    // have an empty batch. The next state will be 'Completed'.
-                    return ODataBatchReaderState.Completed;
-
-                case ODataBatchReaderState.Operation:
-                    // If we find an end boundary in state 'Operation' we have finished
-                    // processing an operation and found the end boundary of either the
-                    // current changeset or the batch.
-                    return this.batchStream.ChangeSetBoundary == null
-                        ? ODataBatchReaderState.Completed
-                        : ODataBatchReaderState.ChangesetEnd;
-
-                case ODataBatchReaderState.ChangesetStart:
-                    // If we find an end boundary when in state 'ChangeSetStart' it means that
-                    // we have an empty changeset. The next state will be 'ChangeSetEnd'
-                    return ODataBatchReaderState.ChangesetEnd;
-
-                case ODataBatchReaderState.ChangesetEnd:
-                    // If we are at the end of a changeset and find an end boundary
-                    // we reached the end of the batch
-                    return ODataBatchReaderState.Completed;
-
-                case ODataBatchReaderState.Completed:
-                    // We should never get here when in Completed state.
-                    throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataBatchReader_GetEndBoundary_Completed));
-
-                case ODataBatchReaderState.Exception:
-                    // We should never get here when in Exception state.
-                    throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataBatchReader_GetEndBoundary_Exception));
-
-                default:
-                    // Invalid enum value
-                    throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataBatchReader_GetEndBoundary_UnknownValue));
+                throw new ODataException(Strings.ODataBatchReader_MaxBatchSizeExceeded(this.inputContext.MessageReaderSettings.MessageQuotas.MaxPartsPerBatch));
             }
         }
 
         /// <summary>
-        /// Reads the next part from the batch message payload.
+        /// Increases the size of the current change set; throws if the allowed limit is exceeded.
         /// </summary>
-        /// <returns>true if more information was read; otherwise false.</returns>
-        private bool ReadSynchronously()
+        protected void IncreaseChangesetSize()
         {
-            return this.ReadImplementation();
+            this.currentChangeSetSize++;

```

Comment:
>currentChangeSetSize [](start = 17, length = 20)

Seems like the validation should be done before the variable is incremented and left in an invalid state.

Reply:
This is the original way of how sizes are incremented in original ODataBatchReader/Writer ([https://github.com/OData/odata.net/blob/master/src/Microsoft.OData.Core/ODataBatchReader.cs#L763](https://github.com/OData/odata.net/blob/master/src/Microsoft.OData.Core/ODataBatchReader.cs#L763)).
When the changeset size exceeds the max allowed, the processing of the changeset has to be aborted and the changeset size will always be reset anyway.
Good observation though. There shouldn't be data-racing issue, therefore I guess we can change it if you have strong opinion on this.

---
In reply to: [140842919](https://github.com/OData/odata.net/pull/960#discussion_r140842919) [](ancestors = 140842919)